include/ginkgo/extensions/kokkos/spaces.hpp

```diff
@@ -0,0 +1,284 @@
+struct compatible_space<MemorySpace, OmpExecutor>
+    : compatible_space<MemorySpace, ReferenceExecutor> {};
+#endif
+
+#ifdef KOKKOS_ENABLE_CUDA
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, CudaExecutor> {
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::Cuda, MemorySpace>::accessible;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_HIP
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, HipExecutor> {
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::HIP, MemorySpace>::accessible;
+};
+#endif
+
+#ifdef KOKKOS_ENABLE_SYCL
+template <typename MemorySpace>
+struct compatible_space<MemorySpace, DpcppExecutor> {
+    static constexpr bool value =
+        Kokkos::SpaceAccessibility<Kokkos::Experimental::SYCL,
+                                   MemorySpace>::accessible;
+};
+#endif
+
+
+/**
+ * Checks if the memory space is accessible by the executor
+ *
+ * @tparam MemorySpace  A Kokkos memory space type
+ * @tparam ExecType  A Ginkgo executor type
+ *
+ * @return  true if the memory space is accessible by the executor
+ */
+template <typename MemorySpace, typename ExecType>
+inline bool check_compatibility(MemorySpace, std::shared_ptr<const ExecType>)
+{
+    return compatible_space<MemorySpace, ExecType>::value;
+}
+
+
+/**
+ * Checks if the memory space is accessible by the executor
+ *
+ * @tparam MemorySpace  A Kokkos memory space
+ * @param exec  The executor which should access the memory space
+ *
+ * @return  true if the memory space is accessible by the executor
+ */
+
+template <typename MemorySpace>
+inline bool check_compatibility(MemorySpace space,
+                                std::shared_ptr<const Executor> exec)
+{
+    if (auto p = std::dynamic_pointer_cast<const ReferenceExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    if (auto p = std::dynamic_pointer_cast<const OmpExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    if (auto p = std::dynamic_pointer_cast<const CudaExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    if (auto p = std::dynamic_pointer_cast<const HipExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    if (auto p = std::dynamic_pointer_cast<const DpcppExecutor>(exec)) {
+        return check_compatibility(space, p);
+    }
+    GKO_NOT_IMPLEMENTED;
+}
+
+
+/**
+ * Throws if the memory space is ~not~ accessible by the executor associated
+ * with the passed in Ginkgo object.
+ *
+ * @tparam MemorySpace  A Kokkos memory space type.
+ * @tparam T  A Ginkgo type that has the member function `get_executor`.
+ *
+ * @param obj  Object which executor is used to check the access to the memory
+ * space.
+ * @param space  The Kokkos memory space to compare against.
+ */
+template <typename MemorySpace, typename T>
+inline void assert_compatibility(T&& obj)
+{
+    if (!check_compatibility(space, obj.get_executor())) {
```

Comment:
no space input.
with the change, maybe you also need to change the check_compatibility without space.